File: watch-apple/Sources-iOS/UploadProgressCard.swift
import SwiftUI

// Prominente Live-Upload-Karte (Home + Sessions): eigene Sessions im Zwischenzustand
// (recording/live), sobald Chunks am Server ankommen — inkl. „GPS da"-Anzeige, Stall-Hinweis
// (>5 min kein Chunk) und Tap -> Detailseite (dort triggert der Server die gps_only-Analyse).
// Pollt schnell (4 s) solange etwas läuft, sonst träge (20 s). Rendert nichts, wenn leer.
// Parität zur PWA/Android. NICHT in Community einbauen.
struct UploadProgressCard: View {
    @State private var rows: [InProgressSession] = []

    var body: some View {
        Group {
            if !rows.isEmpty {
                VStack(alignment: .leading, spacing: 8) {
                    // UEBERHOLTE Aufnahmen (die Uhr hat ihren Puffer weitergedreht, es kommt
                    // nichts mehr) stehen NICHT als grosse Karte ueber allem — dort saesse sonst
                    // dauerhaft ein langer Hinweis, obwohl nichts laeuft und nichts zu tun ist,
                    // solange man nicht hineingeht (Jan, 11.09.2026; in der PWA seit demselben
                    // Tag so). Ganz verschwinden duerfen sie auch nicht: eine `live`-Session
                    // steht in der Liste nur unter „Aussortiert", und genau daran lag der Befund
                    // vom 03.09. — 30 haengende Uploads sahen ihre Besitzer NIRGENDS. Deshalb
                    // eine schmale Zeile mit dem Weg zur Session; entschieden wird dort.
                    ForEach(rows.filter { $0.ueberholt != true }) { s in
                        NavigationLink { SessionDetailView(id: s.id) } label: { UploadCardRow(s: s) }
                            .buttonStyle(.plain)
                    }
                    ForEach(rows.filter { $0.ueberholt == true }) { s in
                        NavigationLink { SessionDetailView(id: s.id) } label: { SchmaleZeile() }
                            .buttonStyle(.plain)
                    }
                }
            }
        }
        .task { await poll() }
    }

    private func poll() async {
        while !Task.isCancelled {
            if let r = try? await Api.inProgress() { rows = r }
            let secs: UInt64 = rows.isEmpty ? 20 : 4
            try? await Task.sleep(nanoseconds: secs * 1_000_000_000)
        }
    }
}

/// Eine ueberholte Aufnahme in der Uebersicht: eine Zeile, kein Block. Gegenstueck zu
/// `SchmaleZeile` in der PWA und in der Android-App.
private struct SchmaleZeile: View {
    @AppStorage("appLang") private var lang = "de"

    var body: some View {
        HStack(spacing: 8) {
            Image(systemName: "info.circle").foregroundStyle(.secondary)
            Text(Loc.t("upload.supersededShort", lang))
                .font(.callout).foregroundStyle(.secondary)
                .lineLimit(1).truncationMode(.tail)
            Spacer(minLength: 4)
            Image(systemName: "chevron.right").font(.footnote).foregroundStyle(.tertiary)
        }
        .padding(.horizontal, 12).padding(.vertical, 8)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 10))
    }
}

// Eine Upload-Zeile — von der Liste (Home/Sessions) und von der Detailseite derselben Session
// benutzt, damit beide waehrend eines Uploads genau dasselbe zeigen (Jan, 01.09.).
struct UploadCardRow: View {
    let s: InProgressSession
    @AppStorage("appLang") private var lang = "de"

    var body: some View { rowView(s) }

    private func stalled(_ s: InProgressSession) -> Bool {
        guard let str = s.last_received_at else { return false }
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let d = f.date(from: str) ?? {
            let g = ISO8601DateFormatter(); g.formatOptions = [.withInternetDateTime]; return g.date(from: str)
        }()
        guard let d else { return false }
        return Date().timeIntervalSince(d) > 300
    }

    @ViewBuilder private func rowView(_ s: InProgressSession) -> some View {
        let pct: Double? = (s.upload_total ?? 0) > 0
            ? min(1.0, Double(s.upload_received) / Double(s.upload_total!)) : nil
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 8) {
                Image(systemName: "icloud.and.arrow.up").foregroundColor(.cyan)
                Text(Loc.t("upload.title", lang)).font(.subheadline).bold()
                if let dev = s.device_label, !dev.isEmpty {
                    Text(dev).font(.caption).foregroundColor(.secondary)
                }
                Spacer()
            }
            HStack(spacing: 10) {
                if s.has_gps {
                    HStack(spacing: 3) {
                        Image(systemName: "location.fill").font(.caption)
                        Text(Loc.t("upload.gpsReady", lang))
                        Image(systemName: "checkmark").font(.caption)
                    }.foregroundColor(.cyan).font(.subheadline)
                } else {
                    Text(Loc.t("upload.waiting", lang)).font(.subheadline).foregroundColor(.secondary)
                }
                Text(pct != nil
                     ? "\(Int(pct! * 100)) % · \(s.upload_received)/\(s.upload_total!)"
                     : Loc.t("upload.chunks", lang).replacingOccurrences(of: "{n}", with: "\(s.upload_received)"))
                    .font(.subheadline)
            }
            if let pct { ProgressView(value: pct).tint(.cyan) }
            else { ProgressView().progressViewStyle(.linear).tint(.cyan) }
            // „Ueberholt": seither ist eine neuere Session komplett angekommen, die Uhr hat ihren
            // Puffer also weitergedreht — „App auf der Uhr oeffnen" waere ein falscher Rat.
            if s.ueberholt == true {
                Label(Loc.t(s.has_gps ? "upload.supersededHint" : "upload.supersededEmpty", lang),
                      systemImage: "info.circle")
                    .font(.subheadline).foregroundColor(.orange)
            } else if stalled(s) {
                Label(Loc.t("upload.stalledHint", lang), systemImage: "info.circle")
                    .font(.subheadline).foregroundColor(.orange)
            } else {
                Text(Loc.t("upload.hint", lang)).font(.subheadline).foregroundColor(.secondary)
            }
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color.cyan.opacity(0.10))
        .clipShape(RoundedRectangle(cornerRadius: 14))
    }
}

// Dieselbe Karte auf der Detailseite EINER Session — ohne Tap-Ziel (sie wuerde auf sich selbst
// fuehren). Verschwindet von selbst, sobald der Upload durch ist: dann steht die Session nicht
// mehr in /in-progress und nur noch die Analyse laeuft.
struct SessionUploadCard: View {
    let id: Int
    @State private var row: InProgressSession? = nil

    var body: some View {
        Group {
            if let row { UploadCardRow(s: row) }
        }
        .task { await poll() }
    }

    private func poll() async {
        while !Task.isCancelled {
            if let r = try? await Api.inProgress() { row = r.first { $0.id == id } }
            let secs: UInt64 = row == nil ? 20 : 4
            try? await Task.sleep(nanoseconds: secs * 1_000_000_000)
        }
    }
}

File: android/app/src/main/java/org/pumpfoil/app/UploadProgressCard.kt
package org.pumpfoil.app

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.CloudUpload
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.LocationOn
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay

// Prominente Live-Upload-Karte (Home + Sessions): eigene Sessions im Zwischenzustand
// (recording/live), sobald Chunks am Server ankommen — inkl. „GPS da"-Anzeige, Stall-Hinweis
// (>5 min kein Chunk) und Klick -> Detailseite (dort triggert der Server die gps_only-Analyse).
// Pollt schnell (4 s) solange etwas läuft, sonst träge (20 s). Rendert nichts, wenn leer.
// Parität zur PWA. Farben aus dem Theme (kippt automatisch light/dark).
@Composable
fun UploadProgressCard(onOpen: (Int) -> Unit, modifier: Modifier = Modifier) {
    var rows by remember { mutableStateOf<List<InProgressSession>>(emptyList()) }
    LaunchedEffect(Unit) {
        while (true) {
            rows = try { Api.inProgress() } catch (_: Exception) { rows }
            delay(if (rows.isNotEmpty()) 4000L else 20000L)
        }
    }
    if (rows.isEmpty()) return
    // Column (mit Bottom-Abstand) nur wenn gerendert -> kein Phantom-Gap, wenn leer.
    Column(modifier.padding(bottom = 10.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
        rows.forEach { s -> UploadRow(s, onOpen) }
    }
}

// Dieselbe Karte auf der Detailseite EINER Session (Jan, 01.09.: „die session sollte auch die
// gleiche anzeige verwenden wie die session-uebersicht waehrend eines uploads"). Ohne Klickziel —
// sie wuerde auf sich selbst verlinken. Rendert nichts, sobald der Upload durch ist (dann steht
// die Session nicht mehr in /in-progress) und die Analyse allein weiterlaeuft.
@Composable
fun SessionUploadCard(id: Int, modifier: Modifier = Modifier) {
    var row by remember(id) { mutableStateOf<InProgressSession?>(null) }
    LaunchedEffect(id) {
        while (true) {
            row = try { Api.inProgress().firstOrNull { it.id == id } } catch (_: Exception) { row }
            delay(if (row != null) 4000L else 20000L)
        }
    }
    row?.let { UploadRow(it, onOpen = null, modifier = modifier) }
}

@Composable
private fun UploadRow(s: InProgressSession, onOpen: ((Int) -> Unit)?, modifier: Modifier = Modifier) {
    val accent = MaterialTheme.colorScheme.primary
    val pct = s.uploadTotal?.takeIf { it > 0 }?.let {
        (s.uploadReceived.toFloat() / it).coerceIn(0f, 1f)
    }
    val stalled = remember(s.lastReceivedAt) {
        s.lastReceivedAt?.let {
            try {
                System.currentTimeMillis() -
                    java.time.OffsetDateTime.parse(it).toInstant().toEpochMilli() > 5 * 60 * 1000
            } catch (_: Exception) { false }
        } ?: false
    }
    val inhalt: @Composable () -> Unit = {
        Column(Modifier.padding(14.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(Icons.Filled.CloudUpload, null, tint = accent, modifier = Modifier.size(20.dp))
                Spacer(Modifier.width(8.dp))
                Text(I18n.t("upload.title"), fontWeight = FontWeight.SemiBold,
                    style = MaterialTheme.typography.titleSmall)
                if (!s.deviceLabel.isNullOrBlank()) {
                    Spacer(Modifier.width(8.dp))
                    Text(s.deviceLabel, style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
            }
            Spacer(Modifier.height(6.dp))
            Row(verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                if (s.hasGps) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Icon(Icons.Filled.LocationOn, null, tint = accent, modifier = Modifier.size(16.dp))
                        Spacer(Modifier.width(3.dp))
                        Text(I18n.t("upload.gpsReady"), color = accent,
                            style = MaterialTheme.typography.bodyMedium, fontWeight = FontWeight.Medium)
                        Icon(Icons.Filled.CheckCircle, null, tint = accent, modifier = Modifier.size(14.dp))
                    }
                } else {
                    Text(I18n.t("upload.waiting"), style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant)
                }
                Text(
                    if (pct != null) "${(pct * 100).toInt()} % · ${s.uploadReceived}/${s.uploadTotal}"
                    else I18n.t("upload.chunks").replace("{n}", s.uploadReceived.toString()),
                    style = MaterialTheme.typography.bodyMedium,
                )
            }
            Spacer(Modifier.height(8.dp))
            if (pct != null) {
                LinearProgressIndicator(progress = { pct }, modifier = Modifier.fillMaxWidth().height(6.dp),
                    color = accent)
            } else {
                LinearProgressIndicator(modifier = Modifier.fillMaxWidth().height(6.dp), color = accent)
            }
            Spacer(Modifier.height(8.dp))
            // „Ueberholt": seither ist eine neuere Session komplett angekommen, die Uhr hat ihren
            // Puffer also weitergedreht — „App auf der Uhr oeffnen" waere ein falscher Rat.
            if (s.ueberholt) {
                Row(verticalAlignment = Alignment.Top) {
                    Icon(Icons.Filled.Info, null, tint = MaterialTheme.colorScheme.tertiary,
                        modifier = Modifier.size(16.dp))
                    Spacer(Modifier.width(6.dp))
                    Text(I18n.t(if (s.hasGps) "upload.supersededHint" else "upload.supersededEmpty"),
                        style = MaterialTheme.typography.bodyMedium)
                }
            } else if (stalled) {
                Row(verticalAlignment = Alignment.Top) {
                    Icon(Icons.Filled.Info, null, tint = MaterialTheme.colorScheme.tertiary,
                        modifier = Modifier.size(16.dp))
                    Spacer(Modifier.width(6.dp))
                    Text(I18n.t("upload.stalledHint"), style = MaterialTheme.typography.bodyMedium)
                }
            } else {
                Text(I18n.t("upload.hint"), style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
        }
    }
    val farben = CardDefaults.cardColors(containerColor = accent.copy(alpha = 0.10f))
    val form = RoundedCornerShape(14.dp)
    if (onOpen != null) {
        Card(onClick = { onOpen(s.id) }, modifier = modifier.fillMaxWidth(), colors = farben, shape = form) { inhalt() }
    } else {
        Card(modifier = modifier.fillMaxWidth(), colors = farben, shape = form) { inhalt() }
    }
}
